Use tidy-html5 to validate the term-table page

octopusinvitro opened the pull request:
# What does this do?

HTML-validates the term-table page and in doing so addresses [this comments](https://github.com/everypolitician/viewer-sinatra/pull/15433#discussion_r80199278)

# Why was this needed?

HTML was broken

# Relevant Issue(s)

https://github.com/everypolitician/everypolitician/issues/505

# Implementation notes

None

# Screenshots

None

# Notes to Reviewer

Errors corrected:
- An "img" element must have an "alt" attribute, except under certain
conditions. For details, consult guidance on providing text alternatives
for images. (All occurrences)

# Notes to Merger

This branch has been rebased with master. An HTML helper is being introduced as part of https://github.com/everypolitician/viewer-sinatra/pull/15584, a helper to which this PR would be adding new methods, so that PR should probably be merged first so that adding a new helper is not part of this PR, but just modifying it.

tmtmtmtm commented on t/web/term_table/phillipines.rb:
```diff
@@ -0,0 +1,38 @@
+require 'test_helper'
+require_relative '../../../app'
+
+describe 'Phillipines' do
```
Typo: "Philippines"

tmtmtmtm commented on app.rb:
```diff
@@ -12,6 +12,7 @@
 require 'everypolitician/popolo'
 
 require_relative './lib/popolo_helper'
+require_relative './helpers/html'
```
Shouldn't we be consistent where the helpers live? If we're making a new 'helpers' directory for them, then we should probably move the `popolo_helper` there too.

octopusinvitro replied:
I thought about that and I think it may be a good idea. I've seen that in other frameworks as Rails or Middleman and I think it helps to keep concerns separated. But I am happy to put this file under lib if we prefer it so.

tmtmtmtm commented on helpers/html.rb:
```diff
@@ -0,0 +1,12 @@
+    end
+  end
+
+  helpers HTMLEscapeHelper
```
With the `PopoloHelper` we [set it up](https://github.com/everypolitician/viewer-sinatra/blob/ee9d1dd664f4094a85e9ba7b38338213fb14b570/app.rb#L18) in `app.rb` itself. I'm not sure whether it's generally preferred to do it that way or this, but again I think we should be consistent.

octopusinvitro replied:
See previous comment. I am happy to move it to app.rb though.

tmtmtmtm commented on t/web/term_table/phillipines.rb:
```diff
@@ -0,0 +1,38 @@
+
+    it 'doesnt break for names with double quotes' do
+      img = tiangco.css('img.person-card__image')
+      img.attr('alt').text.must_equal 'Member headshot for Tiangco, Tobias "Toby" M.'
```
There are actually two different images with this text, and only one of them is getting checked here. Both of them actually get returned from this css selector, though, so it should be quite easy to check both.

tmtmtmtm commented on helpers/html.rb:
```diff
@@ -0,0 +1,12 @@
+
+module Sinatra
+  module HTMLEscapeHelper
+    def h(text)
```
Is `h` a general standard for naming functions like this? I'm generally a fan of brevity, but I'm a little hesitant here, unless this is a universally recognised approach.

octopusinvitro replied:
Ah, this was copied from [the official documentation](http://www.sinatrarb.com/extensions.html#extending-the-request-context-with-sinatrahelpers), I can certainly rename it

octopusinvitro commented:
@tmtmtmtm Feedback addressed in fixup commits

* I tried to call the HTML helper from the app but it couldn't make it work. So I just moved the helper from the helper folder to the lib folder. I also renamed the `h` method to `escape_html`.
* I downloaded and used the popolo file from the Bahamas and removed the one for the Philippines.
* I removed the test file for the Philippines and added a new one for the Bahamas. The switch to the Bahamas file allowed me to discover another bug, when one of the urls in the sources of the data contains spaces. I fixed that by adding another method to the `HTMLEscapeHelper` called `escape_uri` which uses `URI.escape`.

tmtmtmtm commented on views/term_table.erb:
```diff
@@ -198,7 +198,7 @@
         <div class="container">
           <% if @page.data_sources %>
-            <p>Main Source<% if @page.data_sources.size > 1 %>s<% end %>: <%= @page.data_sources.map { |url| %Q(<a href="#{url}">#{url}</a>) }.join(", ") %></p>
+            <p>Main Source<% if @page.data_sources.size > 1 %>s<% end %>: <%= @page.data_sources.map { |url| %Q(<a href="#{escape_uri(url)}">#{url}</a>) }.join(", ") %></p>
```
This approach concerns me. If the problem is that EP is giving us invalid URLs, then we should fix that upstream, so that everyone consuming the data doesn't need to do things like this. 

Whichever approach we take, however, this feels like something that deserves a PR of its own, properly explaining the problem and solution etc., rather than cramming more into _this_ PR.

tmtmtmtm commented:
@octopusinvitro when you say "it couldn't make it work", what was the problem?

octopusinvitro commented:
@tmtmtmtm This branch has been rebased with master. An HTML helper is being introduced as part of https://github.com/everypolitician/viewer-sinatra/pull/15584, a helper to which this PR would be adding new methods, so that PR should probably be merged first so that adding a new helper is not part of this PR, but just modifying it.

octopusinvitro commented:
Closed as it was moved to #15587
